### PR TITLE
[tt-train] Softmax operation for tt-train

### PR DESCRIPTION
**Pull request by vmelnykovTT**
### Problem description
We need a new implementation of the softmax operation that supports fp32_dest_acc_en mode and handles cases where the input data does not fit entirely into L1 memory

### What's changed
- Added a new implementation of the softmax operation that meets our requirements, including support for fp32_dest_acc_en mode and handling inputs that exceed L1 memory capacity.
- Fixed a bug in the cross_entropy_bw kernel where the maximum value used for numerical stability was not correctly computed.
- Changed softmax op in scale_dot_product_attention.cpp

### Comparison the result with ttnn_fixed::softmax and ttnn::moreh_softmax
Summarize the changes made and its impact.

The following plots illustrate the impact of different softmax implementations on numerical stability, accuracy, and performance.
I tested 1000 randomized softmax inputs using various approaches. The baseline is a high-precision xt::xarray<float> implementation.
The goal is to evaluate how closely each approach matches the baseline and how they perform in realistic training scenarios

1. Softmax Output Sum
This plot shows the sum of softmax outputs per test case. Theoretically, softmax output should sum to 1 — which is clearly observed in the baseline.
The plot highlights deviations for ttnn_fixed and the new softmax op. The new op demonstrates improved consistency compared to ttnn_fixed.
sum_plot_ttnn_fixed

2. MSE Compared to Baseline
This plot compares the Mean Squared Error (MSE) of softmax outputs against the float baseline.
Two comparisons are shown:
	•	New op vs ttnn_fixed::softmax
	•	New op vs ttnn::moreh_softmax
mse_plot_ttnn_fixed
mse_plot_moreh

3. Training Loss Over Steps
This plot shows training loss progression after 5000 steps using GPT2s.
Unlike ttnn_fixed::softmax, the new op does not exhibit a loss spike near the end, indicating improved training stability.
losses_gpt2s_full_training

4. Step Time Comparison
This plot compares step time (in ms) during nano_gpt training.
While the new op is slightly slower than ttnn_fixed::softmax, it provides better numerical accuracy and stability — a tradeoff worth considering in many cases.

Mean Step Time for new op: 268.64 ms
Mean Step Time for ttnn_fixed::softmax: 266.36 ms
step_time_nano_gpt

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/tests/ops/softmax_test.cpp [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    std::cout << "Expected Result:\n";
+    expected_result_print.print();
+
+    // Check if the result is close to the expected result
+    auto result_xtensor = core::to_xtensor(result);
+    assert((result_xtensor.shape() == expected_result.shape()));
+    EXPECT_TRUE(xt::allclose(result_xtensor, expected_result, 3e-2F, 1e-2F));
+}
+
+// The following function writes results to a file.
```
Comment (rfurko-tt) [inline tt-train/tests/ops/softmax_test.cpp]: lets remove this before merge, you can save it locally

File: tt-train/sources/ttml/metal/ops/softmax/device/softmax_device_operation.cpp [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    if (preallocated_output_tensor.has_value()) {
+        check_tensor(
+            preallocated_output_tensor.value(),
+            "Preallocated Output",
+            tt::tt_metal::Layout::TILE,
+            tt::tt_metal::DataType::BFLOAT16);
+    }
+
+    // Validate the dimension argument
+    uint32_t input_rank = static_cast<int32_t>(input_tensor.logical_shape().rank());
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/softmax/device/softmax_device_operation.cpp]: uint32_t or int32_t
not clear what's your plan here

File: tt-train/sources/ttml/metal/ops/softmax/device/kernels/dataflow/reader_softmax_interleaved_start_id.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <dataflow_api_addrgen.h>
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/softmax/device/kernels/dataflow/reader_softmax_interleaved_start_id.cpp]: are you sure you need this two includes?

**Review by rfurko-tt — APPROVED** [review]
LGTM

Well done @vmelnykovTT!

File: tt-train/sources/ttml/ops/scaled_dot_product_attention.cpp [hunk 4]
```diff
@@ -178,7 +179,8 @@ autograd::TensorPtr scaled_dot_product_attention(
             false);
     }
     // (B, H, S, S)
-    auto attention_weights = ttnn_fixed::softmax(qk_scaled, /* axis */ 3);
+    // auto attention_weights = ttnn_fixed::softmax(qk_scaled, /* axis */ 3);
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/ops/scaled_dot_product_attention.cpp]: remove comment

File: tt-train/sources/ttml/metal/ops/softmax/device/softmax_device_operation.cpp [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    // Validate the dimension argument
+    int32_t input_rank = static_cast<int32_t>(input_tensor.logical_shape().rank());
+    // TT_FATAL(
+    //     (args.dim < input_rank) && (-input_rank <= args.dim),
+    //     "`dim` must be in the range [-input_rank, input_rank) (provided: dim = {}, rank = {}).",
+    //     args.dim,
+    //     input_rank);
+
+    TT_FATAL(
+        args.dim == (input_rank - 1U) || args.dim == -1U,
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/metal/ops/softmax/device/softmax_device_operation.cpp]: Substitute 1U -> 1 in both places
